testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/OfflineTokenTest.java

```diff
@@ -255,24 +256,24 @@ public void offlineTokenBrowserFlow() throws Exception {
         setTimeOffset(3000000);
 
         OAuthClient.AccessTokenResponse response = oauth.doRefreshTokenRequest(newRefreshTokenString, "secret1");
+        RefreshToken newRefreshToken = oauth.parseRefreshToken(newRefreshTokenString);
         Assert.assertEquals(400, response.getStatusCode());
         assertEquals("invalid_grant", response.getError());
 
-        events.expectRefresh(offlineToken.getId(), sessionId)
+        events.expectRefresh(offlineToken.getId(), newRefreshToken.getSessionState())
```

Comment:
Maybe will be slightly better to test with `offlineToken.getSessionState()` instead of `newRefreshToken.getSessionState()` ? Or keep it like this, but add another line with `Assert.assertEquals(offlineToken.getSessionState(), newRefreshToken.getSessionState())` .

I understand that the "sessionId" in the offline token might not be the same as the ID of "online" session. However I think that "sessionId" of offline token should not change during each offline token refresh? Hence it may be good to test that sessionState of the old offline token corresponds to the sessionState in the new offline token, which corresponds to the ID of the offline session.

Reply:
I checked this and I found out that `sessionState` is different after the offline token refresh when using `map` implementation.
Reason is in `TokenManager.initToken` where `sessionState` is set like `token.setSessionState(session.getId());` Initial token /offlineToken was created using the online user session thus there's a different session Id. The following offline refresh operates with offline user session (online is already expired) thus there's is offline user session id set as `sessionState`. 
Do you see something wrong here?

Reply:
@martin-kanis I see. Thanks for the explanation.

To me, it would look more logical that all the offline tokens use the sessionId of the offline session rather than sessionId of the "online" session. So my vote is, that all the tokens issued during the codeToToken request (the request, which creates offline session) use offline sessionId (including the accessToken and IDToken issued during this request). @hmlnarik WDYT?

Reply:
I agree with @mposolda. If turns out to be not be straightforward, I'm open to delivering this in a separate JIRA